NeuroID/src/main/java/com/neuroid/tracker/service/NIDSamplingService.kt

```diff
@@ -0,0 +1,85 @@
+package com.neuroid.tracker.service
+
+import com.neuroid.tracker.utils.NIDLogWrapper
+import com.neuroid.tracker.utils.RandomGenerator
+
+class NIDSamplingService(
+    val logger: NIDLogWrapper,
+    // we need this to be replaceable for the tests!
+    var randomGenerator: RandomGenerator,
+    var configService: NIDRemoteConfigService,
+) {
+    private var isSessionFlowSampled = true
+
+    /**
+     * given a site id, tell me if we sample events or not. the site ID will be compared
+     */
+    fun updateIsSampledStatus(siteID: String?) {
+        val currentSampleRate = retrieveSampleRate(siteID)
+        if (currentSampleRate >= MAX_SAMPLE_RATE) {
+            logger.d(msg="updateIsSampledStatus() we are sampling at max or > than max ($currentSampleRate)")
+            isSessionFlowSampled = true
+            return
+        }
+        val randomValue = randomGenerator.getRandom(MAX_SAMPLE_RATE)
+        logger.d(msg="updateIsSampledStatus() random value: $randomValue sampleRate: $currentSampleRate")
+        if (randomValue < currentSampleRate) {
+            isSessionFlowSampled = true
+            logger.d(msg="updateIsSampledStatus() we are sampling")
+            return
+        }
+
+        logger.d(msg="updateIsSampledStatus() we are not sampling")
+        isSessionFlowSampled = false
+    }
+
+
+    /**
+     * Return the sample rate, 0 indicates sample everything, 100 indicates sample nothing!
+     */
+    private fun retrieveSampleRate(siteID: String?): Int {
+        // is the site id passed in a linked site id in the options?, if so return the linked site
+        // id sample rate
+        if (isLinkedSiteID(siteID)) {
+            configService.getRemoteNIDConfig().linkedSiteOptions[siteID]?.let {
+                logger.d(msg = "retrieveSampleRate site id is in options, ${it.sampleRate}")
+                return it.sampleRate
```

Comment:
What if the siteID is in the linkedSiteOptions but the key `sampleRate` is not?
I.e. 
```
{
  "linkedSiteOptions:{
     "site1": { "otherKey":"foo" }
   }
}
```

will this fall back to the DEFAULT_SAMPLE_RATE or will it return undefined/blow up?

Reply:
added default checks here
`return it.sampleRate ?: DEFAULT_SAMPLE_RATE`

Reply:
As we define more options here, we will need to add It to the NIDLinkedSiteOption data class.